File: clients/android/app/src/main/java/com/familyconnect/app/ControlProfiles.java
package com.familyconnect.app;

import com.google.gson.*;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.regex.*;
import static com.familyconnect.app.ControlJson.*;

/** Schema2 profiles, distinct from Android's legacy import parser. Never resolves names. */
final class ControlProfiles {
    private static final List<String> AWG=Arrays.asList("Jc Jmin Jmax S1 S2 S3 S4 H1 H2 H3 H4 I1 I2 I3 I4 I5".split(" "));
    private static String s(JsonObject v,String key){return text(v.get(key));}
    private static void id(String v){require(v.matches("[a-zA-Z0-9_-]{1,64}"));}
    private static InetAddress ip(String value)throws Exception {
        require(value.matches("[0-9a-fA-F:.]+"));
        if(!value.contains(":")) {
            String[] parts=value.split("\\.",-1);require(parts.length==4);
            for(String p:parts){require(p.matches("0|[1-9][0-9]{0,2}"));require(Integer.parseInt(p)<=255);}
        }
        return InetAddress.getByName(value);
    }
    private static String canonical(InetAddress ip) {
        byte[] bytes=ip.getAddress();if(bytes.length==4)return ip.getHostAddress();
        int[] words=new int[8];for(int i=0;i<8;i++)words[i]=((bytes[2*i]&255)<<8)|(bytes[2*i+1]&255);
        int best=-1,length=1;
        for(int i=0;i<8;){if(words[i]!=0){i++;continue;}int end=i;while(end<8&&words[end]==0)end++;if(end-i>length){best=i;length=end-i;}i=end;}
        StringBuilder result=new StringBuilder();
        for(int i=0;i<8;i++) {
            if(i==best){result.append("::");i+=length-1;continue;}
            if(result.length()>0 && result.charAt(result.length()-1)!=':')result.append(':');result.append(Integer.toHexString(words[i]));
        }
        return result.toString();
    }
    private static void key(String value,boolean canonical,boolean nonzero) {
        byte[] raw=ControlProtocol.base64(value,canonical);require(raw.length==32);boolean any=false;for(byte b:raw)any|=b!=0;require(!nonzero||any);
    }
    private static long number(String value,long low,long high) {
        require(value.matches("[+-]?[0-9]+"));long result=Long.parseLong(value);require(result>=low&&result<=high);return result;
    }
    private static final class Endpoint {
        final String host;final long port;Endpoint(String host,long port){this.host=host;this.port=port;}
        boolean same(Endpoint other){return other!=null&&host.equals(other.host)&&port==other.port;}
    }
    static void validate(JsonObject state)throws Exception {
        fields(state,"schema_version config_id revision issued_at expires_at recipient audience wireguard_public_key min_client_version previous_config_hash signer_key_id gateways transport_profiles");
        id(s(state,"config_id"));integer(state.get("revision"),1);
        long issued=integer(state.get("issued_at"),1),expires=integer(state.get("expires_at"),1);require(expires>issued&&expires-issued<=86400);
        key(s(state,"wireguard_public_key"),true,true);ControlProtocol.version(s(state,"min_client_version"));
        JsonElement previous=state.get("previous_config_hash");require(previous.isJsonNull()||text(previous).matches("[0-9a-f]{64}"));
        require(state.get("gateways").isJsonArray()&&state.get("transport_profiles").isJsonArray());
        JsonArray gateways=state.getAsJsonArray("gateways"),profiles=state.getAsJsonArray("transport_profiles");
        require(gateways.size()>=1&&gateways.size()<=8&&profiles.size()>=1&&profiles.size()<=8);
        Map<String,Endpoint> endpoints=new HashMap<>();
        for(JsonElement value:gateways) {
            fields(value,"gateway_id endpoint port");JsonObject g=value.getAsJsonObject();id(s(g,"gateway_id"));
            String host=s(g,"endpoint");InetAddress address=ip(host);
            require(canonical(address).equals(host)&&!address.isAnyLocalAddress()&&!address.isLoopbackAddress()&&!address.isMulticastAddress());
            long port=integer(g.get("port"),1);require(port<=65535);
            require(endpoints.put(s(g,"gateway_id"),new Endpoint(host,port))==null);
        }
        Set<String> ids=new HashSet<>();
        for(JsonElement value:profiles) {
            fields(value,"profile_id gateway_id transport transport_version config");JsonObject p=value.getAsJsonObject();
            id(s(p,"profile_id"));id(s(p,"gateway_id"));require(ids.add(s(p,"profile_id")));
            String config=s(p,"config");require(!config.isEmpty()&&config.getBytes(StandardCharsets.UTF_8).length<=16384);
            String transport=s(p,"transport");require(Arrays.asList("wireguard","amneziawg","vless-reality").contains(transport));
            require(s(p,"transport_version").equals(transport.equals("amneziawg")?"2.0":"1"));
            Endpoint endpoint=transport.equals("vless-reality")?tcp(config):wg(config,transport.equals("amneziawg"));
            require(endpoint.same(endpoints.get(s(p,"gateway_id"))));
        }
    }
    private static Endpoint tcp(String config)throws Exception {
        JsonElement value=parse(config.getBytes(StandardCharsets.UTF_8));fields(value,"type server port id public_key server_name short_id");JsonObject p=value.getAsJsonObject();
        require(s(p,"type").equals("vless-reality-v1"));String host=s(p,"server");require(!host.contains(":")&&ip(host) instanceof Inet4Address);
        long port=integer(p.get("port"),1);require(port<=65535);
        String id=s(p,"id");require(UUID.fromString(id).toString().equals(id)&&!id.equals("00000000-0000-0000-0000-000000000000"));
        String key=s(p,"public_key");require(key.matches("[A-Za-z0-9_-]{43}"));key(key.replace('-','+').replace('_','/')+"=",true,true);
        String name=s(p,"server_name");require(name.length()<=253&&name.matches("(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\\.)+[a-z]{2,63}"));
        require(s(p,"short_id").matches("(?:[0-9a-f]{2}){1,8}"));return new Endpoint(host,port);
    }
    private static Endpoint wg(String text,boolean awg)throws Exception {
        require(!text.contains("\0"));Map<String,Map<String,String>> sections=new HashMap<>();Map<String,String> current=null;String section="";
        for(String raw:text.split("\\r\\n|[\\n\\r\\u000b\\f\\u001c-\\u001e\\u0085\\u2028\\u2029]")) {
            String line=raw.split("#",2)[0].trim();if(line.isEmpty())continue;
            if(line.startsWith("[")) {
                require(line.equals("[Interface]")||line.equals("[Peer]"));section=line.substring(1,line.length()-1);current=new HashMap<>();require(sections.put(section,current)==null);continue;
            }
            String[] parts=line.split("=",2);require(current!=null&&parts.length==2);String key=parts[0].trim(),v=parts[1].trim();
            String allowed=section.equals("Interface")?"PrivateKey Address DNS MTU ListenPort":"PublicKey PresharedKey Endpoint AllowedIPs PersistentKeepalive";
            require(Arrays.asList(allowed.split(" ")).contains(key)||(section.equals("Interface")&&awg&&AWG.contains(key)));
            require(!v.isEmpty()&&current.put(key,v)==null);
        }
        require(sections.size()==2);Map<String,String> i=sections.get("Interface"),p=sections.get("Peer");
        require("LOCAL_DEVICE_KEY".equals(i.get("PrivateKey"))&&text.split("LOCAL_DEVICE_KEY",-1).length==2);
        key(p.get("PublicKey"),false,false);if(p.containsKey("PresharedKey"))key(p.get("PresharedKey"),false,false);
        for(String address:i.get("Address").split(",",-1)) {
            String[] bits=address.trim().split("/",-1);require(bits.length<=2);InetAddress a=ip(bits[0]);if(bits.length==2)number(bits[1],0,a instanceof Inet4Address?32:128);
        }
        for(String dns:i.get("DNS").split(",",-1))ip(dns.trim());
        Set<String> routes=new HashSet<>();for(String r:p.get("AllowedIPs").split(",",-1))routes.add(r.trim());require(routes.equals(new HashSet<>(Arrays.asList("0.0.0.0/0","::/0"))));
        String endpoint=p.get("Endpoint");int at=endpoint.lastIndexOf(':');require(at>0);String host=endpoint.substring(0,at).replaceAll("^[\\[\\]]+|[\\[\\]]+$","");ip(host);long port=number(endpoint.substring(at+1),1,65535);
        if(i.containsKey("MTU"))number(i.get("MTU"),1280,1500);if(i.containsKey("ListenPort"))number(i.get("ListenPort"),0,65535);if(p.containsKey("PersistentKeepalive"))number(p.get("PersistentKeepalive"),0,65535);
        if(awg)awg(i);return new Endpoint(host,port);
    }
    private static void awg(Map<String,String> fields) {
        for(String name:AWG.subList(0,7)){String value=fields.get(name);require(value.matches("[0-9]{1,5}"));number(value,0,name.equals("Jc")?12:name.startsWith("J")?1280:256);}
        require(Long.parseLong(fields.get("Jmin"))<=Long.parseLong(fields.get("Jmax")));List<long[]> ranges=new ArrayList<>();
        for(String name:AWG.subList(7,11)) {
            String value=fields.get(name);require(value.matches("[0-9]{1,10}(?:-[0-9]{1,10})?"));String[] parts=value.split("-");long low=number(parts[0],5,4294967295L),high=number(parts[parts.length-1],low,4294967295L);
            for(long[] range:ranges)require(low>range[1]||range[0]>high);ranges.add(new long[]{low,high});
        }
        Pattern packet=Pattern.compile("<(?:b 0x([0-9a-fA-F]+)|(r|rd|rc) ([0-9]{1,4})|(t))>");
        for(String name:AWG.subList(11,16)) {
            String value=fields.get(name);if(value==null)continue;Matcher m=packet.matcher(value);int position=0,size=0;
            while(m.find()){require(m.start()==position);position=m.end();if(m.group(1)!=null){require(m.group(1).length()%2==0);size+=m.group(1).length()/2;}else size+=m.group(2)!=null?Integer.parseInt(m.group(3)):4;}
            require(position==value.length()&&size>=1&&size<=1280);
        }
    }
}

File: clients/windows/Core/ControlProfiles.cs
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using static FamilyConnect.ControlProtocol;

namespace FamilyConnect;

// Schema2 profile validation, kept separate from the legacy activation schemas.
internal static class ControlProfiles
{
    static readonly string[] AwgFields = "Jc Jmin Jmax S1 S2 S3 S4 H1 H2 H3 H4 I1 I2 I3 I4 I5".Split(' ');
    static string S(JsonElement value, string key) => Text(value.GetProperty(key));
    static void Id(string value) => Require(Matches(value, "[a-zA-Z0-9_-]{1,64}"));
    static IPAddress Ip(string value)
    {
        var ip = IPAddress.Parse(value);
        // .NET also accepts abbreviated/octal IPv4 addresses; the Python parser does not.
        if (ip.AddressFamily == AddressFamily.InterNetwork) Require(ip.ToString() == value);
        Require(!value.Contains('%')); return ip;
    }
    static void Key(string value, bool canonical = false, bool nonzero = false)
    {
        var bytes = Base64(value, canonical); Require(bytes.Length == 32 && (!nonzero || bytes.Any(b => b != 0)));
    }
    internal static void Validate(JsonElement state)
    {
        Fields(state, "schema_version config_id revision issued_at expires_at recipient audience wireguard_public_key min_client_version previous_config_hash signer_key_id gateways transport_profiles");
        Id(S(state, "config_id")); Integer(state.GetProperty("revision"), 1);
        var issued = Integer(state.GetProperty("issued_at"), 1); var expires = Integer(state.GetProperty("expires_at"), 1);
        Require(expires > issued && expires-issued <= 86400);
        Key(S(state, "wireguard_public_key"), true, true); VersionValue(S(state, "min_client_version"));
        var previous = state.GetProperty("previous_config_hash"); Require(previous.ValueKind == JsonValueKind.Null || Matches(Text(previous), "[0-9a-f]{64}"));
        var gateways = state.GetProperty("gateways"); var profiles = state.GetProperty("transport_profiles");
        Require(gateways.ValueKind == JsonValueKind.Array && gateways.GetArrayLength() is >=1 and <=8);
        Require(profiles.ValueKind == JsonValueKind.Array && profiles.GetArrayLength() is >=1 and <=8);
        var endpoints = new Dictionary<string, (string, long)>(StringComparer.Ordinal);
        foreach (var g in gateways.EnumerateArray())
        {
            Fields(g, "gateway_id endpoint port"); Id(S(g, "gateway_id"));
            var host = S(g, "endpoint"); var ip = Ip(host); var bytes = ip.GetAddressBytes();
            Require(ip.ToString() == host && !IPAddress.IsLoopback(ip) && !ip.Equals(IPAddress.Any) && !ip.Equals(IPAddress.IPv6Any) && !ip.IsIPv6Multicast && !(bytes.Length == 4 && bytes[0] is >=224 and <=239));
            var port = Integer(g.GetProperty("port"), 1); Require(port <= 65535);
            Require(endpoints.TryAdd(S(g, "gateway_id"), (host, port)));
        }
        var ids = new HashSet<string>(StringComparer.Ordinal);
        foreach (var p in profiles.EnumerateArray())
        {
            Fields(p, "profile_id gateway_id transport transport_version config");
            Id(S(p, "profile_id")); Id(S(p, "gateway_id")); Require(ids.Add(S(p, "profile_id")));
            var config = S(p, "config"); Require(config.Length > 0 && Encoding.UTF8.GetByteCount(config) <= 16384);
            var transport = S(p, "transport"); var version = S(p, "transport_version");
            Require(transport is "wireguard" or "amneziawg" or "vless-reality");
            Require(version == (transport == "amneziawg" ? "2.0" : "1"));
            var endpoint = transport == "vless-reality" ? Tcp(config) : WireGuard(config, transport == "amneziawg");
            Require(endpoints.TryGetValue(S(p, "gateway_id"), out var expected) && endpoint == expected);
        }
    }
    static (string, long) Tcp(string text)
    {
        var p = Parse(Encoding.UTF8.GetBytes(text)); Fields(p, "type server port id public_key server_name short_id");
        Require(S(p, "type") == "vless-reality-v1"); var host = S(p, "server"); Require(Ip(host).AddressFamily == AddressFamily.InterNetwork);
        var port = Integer(p.GetProperty("port"), 1); Require(port <= 65535);
        Require(Guid.TryParseExact(S(p, "id"), "D", out var id) && id != Guid.Empty && id.ToString() == S(p, "id"));
        var key = S(p, "public_key"); Require(Matches(key, "[A-Za-z0-9_-]{43}"));
        var raw = Base64(key.Replace('-', '+').Replace('_', '/')+"=", true); Require(raw.Length == 32 && raw.Any(b => b != 0));
        var name = S(p, "server_name"); Require(name.Length <= 253 && Matches(name, "(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\\.)+[a-z]{2,63}"));
        Require(Matches(S(p, "short_id"), "(?:[0-9a-f]{2}){1,8}")); return (host, port);
    }
    static long Number(string text, long low, long high)
    {
        Require(long.TryParse(text, System.Globalization.NumberStyles.AllowLeadingSign, System.Globalization.CultureInfo.InvariantCulture, out var value) && value >= low && value <= high); return value;
    }
    static (string, long) WireGuard(string text, bool awg)
    {
        Require(!text.Contains('\0'));
        var sections = new Dictionary<string, Dictionary<string, string>>(StringComparer.Ordinal);
        Dictionary<string, string>? current = null; string section = "";
        foreach (var raw in Regex.Split(text, "\\r\\n|[\\n\\r\\v\\f\\x1c-\\x1e\\x85\\u2028\\u2029]"))
        {
            var line = raw.Split('#', 2)[0].Trim(); if (line.Length == 0) continue;
            if (line.StartsWith('['))
            {
                Require(line is "[Interface]" or "[Peer]"); section = line[1..^1];
                current = new(StringComparer.Ordinal); Require(sections.TryAdd(section, current)); continue;
            }
            var parts = line.Split('=', 2); Require(current != null && parts.Length == 2);
            var key = parts[0].Trim(); var value = parts[1].Trim();
            var allowed = section == "Interface" ? "PrivateKey Address DNS MTU ListenPort" : "PublicKey PresharedKey Endpoint AllowedIPs PersistentKeepalive";
            Require(allowed.Split(' ').Contains(key) || (section == "Interface" && awg && AwgFields.Contains(key)));
            Require(value.Length > 0 && current!.TryAdd(key, value));
        }
        Require(sections.Count == 2); var i = sections["Interface"]; var p = sections["Peer"];
        Require(i["PrivateKey"] == "LOCAL_DEVICE_KEY" && text.Split("LOCAL_DEVICE_KEY").Length == 2);
        Key(p["PublicKey"]); if (p.TryGetValue("PresharedKey", out var psk)) Key(psk);
        foreach (var address in i["Address"].Split(','))
        {
            var bits = address.Trim().Split('/'); Require(bits.Length <= 2); var ip = Ip(bits[0]);
            if (bits.Length == 2) Number(bits[1], 0, ip.AddressFamily == AddressFamily.InterNetwork ? 32 : 128);
        }
        foreach (var dns in i["DNS"].Split(',')) Ip(dns.Trim());
        Require(p["AllowedIPs"].Split(',').Select(x => x.Trim()).ToHashSet().SetEquals(new[]{"0.0.0.0/0", "::/0"}));
        var at = p["Endpoint"].LastIndexOf(':'); Require(at > 0);
        var host = p["Endpoint"][..at].Trim('[', ']'); Ip(host); var port = Number(p["Endpoint"][(at+1)..], 1, 65535);
        if (i.TryGetValue("MTU", out var mtu)) Number(mtu, 1280, 1500);
        if (i.TryGetValue("ListenPort", out var listen)) Number(listen, 0, 65535);
        if (p.TryGetValue("PersistentKeepalive", out var keep)) Number(keep, 0, 65535);
        if (awg) ValidateAwg(i);
        return (host, port);
    }
    static void ValidateAwg(Dictionary<string, string> fields)
    {
        foreach (var name in AwgFields.Take(7))
        {
            var value = fields[name]; Require(Matches(value, "[0-9]{1,5}"));
            Number(value, 0, name == "Jc" ? 12 : name.StartsWith('J') ? 1280 : 256);
        }
        Require(long.Parse(fields["Jmin"]) <= long.Parse(fields["Jmax"]));
        var ranges = new List<(long Low, long High)>();
        foreach (var name in AwgFields.Skip(7).Take(4))
        {
            var value = fields[name]; Require(Matches(value, "[0-9]{1,10}(?:-[0-9]{1,10})?"));
            var parts = value.Split('-'); var low = Number(parts[0], 5, uint.MaxValue); var high = Number(parts[^1], low, uint.MaxValue);
            Require(!ranges.Any(r => low <= r.High && r.Low <= high)); ranges.Add((low, high));
        }
        foreach (var name in AwgFields.Skip(11))
        {
            if (!fields.TryGetValue(name, out var value)) continue;
            int position = 0, size = 0;
            foreach (Match m in Regex.Matches(value, "<(?:b 0x([0-9a-fA-F]+)|(r|rd|rc) ([0-9]{1,4})|(t))>"))
            {
                Require(m.Index == position);
                if (m.Groups[1].Success) { Require(m.Groups[1].Length % 2 == 0); size += m.Groups[1].Length / 2; }
                else size += m.Groups[2].Success ? int.Parse(m.Groups[3].Value) : 4;
                position += m.Length;
            }
            Require(position == value.Length && size is >=1 and <=1280);
        }
    }
}
